pkg/engine/engine.go

```diff
@@ -239,8 +240,17 @@ func (e *Engine) QueueRun(request *api.RunRequest, sources *api.UnpackedSources)
 				Created: time.Now().UTC(),
 			},
 		},
-		CreatedBy: task.CreatedBy(request.CreatedBy),
-	})
+		CreatedBy: cby,
+	}
+
+	// Remove existing tasks from same branch end repo before pushing a new task
+	err := e.queue.RemoveExisting(newTask.CreatedBy.Branch, newTask.CreatedBy.Repo)
+
+	if err != nil {
+		return "", err
+	}
+
+	err = e.queue.Push(newTask)
```

Comment:
There is a race condition + it's quite stateful,
What about we implement a single operation? (`PushWithoutDuplicate`, `PushUniqTask`, or some better naming).

Reply:
I'm not sure I follow - you mean replacing the lines 247 - 253 with a single operation, e.g. `PushUniqTask` ?

Reply:
Yes, one way to reason about this:
- Both operations lock the same resource.
- The state at the end of Push is correct if, and only if, we called it "right after" RemoveExisting.

That's a smell that:
- if you use one part of the API (Push) without the other (RemoveExisting), you might inadvertently create bugs in the app,
- interleaving the calls might trigger a race condition
  - Thread1: RemoveExisting(A, B)
  - Thread2: RemoveExisting(A, B)
  - Thread1: Push(A, B)
  - Thread2: Push(A, B)
  - => duplicate output.

Reply:
I see, thanks for the write-up. I wanted to avoid changing the Queue API, but I see that adding a new unified operation is necessary.